### DISCUSSION TITLE
Add 2D editor tools (like align, distribute and reorder) inspired by 2d design software conventions

### DISCUSSION THREAD

**Discussion by lostminds** [godotengine/godot-proposals#5750] [Editor]
When designing 2D UI scenes you can use containers and set their content alignment and anchoring rules to set up how dynamic resizing should be handled and UI elements should be placed. However, this is quite a complex set up, requiring a lot of steps and knowledge, and in some cases you just quickly want to move nodes to line up correctly in relation to each other within a container. 

To ease on-boarding and improve the workflow for experienced users I'd like to discuss the possibility of adding a couple of common tools available in many 2d design applications. These would not add any new layout systems or constrains on the nodes, just work as shortcuts and alternative ways of editing the already existing node parameters. 

Like in 2d design applications they could appear as icons in the top bar based on your selection in the 2d editor:
![Screen Shot 2022-11-09 at 09 17 14](https://user-images.githubusercontent.com/17763524/200781497-e4951743-608a-4e81-af38-54ebc76c8c11.jpg)

Here are three examples of such tools, and how I think they could be used in Godot (illustration of the corresponding tools in Affinity Designer): 
![Screen Shot 2022-11-09 at 09 20 09](https://user-images.githubusercontent.com/17763524/200782161-b2d9579b-94c3-4b50-bd3b-9ae585ba35f0.jpg)

**Reorder** (Move to front, move forward, move backward, move to back)
A tool to simply move one or more selected nodes backward and forward in the draw order, one step or all the way to the front or back. In many cases this would just be a matter of moving nodes up/down in the node tree (there's a keyboard shortcut for this), but the difference is the addition of the useful action to move all the way to the back or front. It also has the usability benefit of focusing on the display order (front to back) instead of the order in the node tree where it's not obvious if nodes will be draw from the top or from the bottom. 

**Align**
A tool available when you have two or more selected nodes (typically child nodes with the same parent) to let you quickly adjust their position to align them relative to each other. Like centered horizontally to each other or aligning the top edges. This would not change the anchor setting, add any new containers or constraints, just change the transform offset of the nodes. 

**Distribute**
A tool available when you have three or more selected nodes (typically child nodes with the same parent) to let you distribute their positions evenly vertically or horizontally from a selected edge. This would not change the anchor setting, add any new containers or constraints, just change the offset of the middle elements. 

**Option-drag duplicate** (Bonus)
Add support for the convention to quickly duplicate nodes by holding option (maybe some other modifier on linux) and clicking and dragging to duplicated an object and move the new copy instead. Once you're used to this it's very very convenient.

Together I think the addition of some or all of these tools inspired by 2d design software would make a lot of difference to the many users who are experienced with working in 2D design software and use these kinds of tools. And as they are just UI shortcuts, and don't introduce any new constraints or underlying layout system changes I think it would be a good candidate for adding. What do you think about this?

**Comment by Calinou** [comment]
I remember seeing a pull request implementing this feature, but no consensus was reached on implementing it. Also, this may be a duplicate of another proposal that's already open, but I can't find it right now.

**Comment by golddotasksquestions** [comment]
@Calinou 

Are you referring to this proposal: https://github.com/godotengine/godot-proposals/issues/1542 ?
I don't think this is what OP wants. https://github.com/godotengine/godot-proposals/issues/1542 is suggesting just to bring very commonly used style of GUI interface forward and allow to apply it do many nodes simultaneously. It does not change behavior and neither would require any new behavior.

What OP seems to describe ([also according to OPs Reddit post](https://www.reddit.com/r/godot/comments/yph0wp/aligndistribute_tool_and_altdragduplication_in/)) on the other hand, is new behaviour to Godot. It's something you get is pretty much ever graphics editor, be it vector art or bitmap to speed up making designs and fast and quick alignments, regardless which layer/group/element those objects belong to.

First: You select a couple of objects and then do operations on them to **align**, **distribute** them **_relative to each other_**, seemingly completely disregarding the hierarchical structure (so any parent-child relationships).

The important difference here is how these settings of alignment and distribution would be applied _**relative to each other**_!

How this would work in Godot, where everything is so dependent on this parent-child relationship, OP does not answer in detail. For example doing any of these operations could mean forcing those selected objects to suddenly have a new(?) single parent, or they keep their existing parent, but if that parent node is a Container type for example, there is a warning popping up saying this operation will have no effect on this node because it is controlled by a Container ... there are lots of unanswered questions still how this is supposed to work.

But I think the behavior OP desires is something very useful and would greatly improve and speed up and GUI work or even general design workflow in Godot. Anyone who ever used a visual design software with those features can attest.

**Reply by Calinou** [reply]
> Are you referring to this proposal: https://github.com/godotengine/godot-proposals/issues/1542 ?

No, I was referring to another proposal.

**Reply by lostminds** [reply]
Looking at that proposal and the [UIDesignTool](https://github.com/imjp94/UIDesignTool) plugin it references makes me think that perhaps making a plugin for this functionality would be possible as a first step to try it out. And the plugin seems like a nice usability improvement for working with Text-related controls.

**Comment by lostminds** [comment]
After seeing some reactions I now think that a lot of people misunderstand this since they assume the (probably recommended) workflow of doing all UI layout using containers with layout rules. If doing this correctly it's like drawing a fully constrained CAD sketch and there will be no freedoms left, and no need for the types of tools I propose.

While very powerful, this workflow is however quite clunky and cumbersome to work with, and there is the alternative more freeform method of working with Anchors instead. Basically giving UI elements a reference point (like the top left corner of their parent container) but then allowing you to position them freely. This is how I've worked with UI layouts since it's more like how I'm used to working. And in this situation having the type of align and distribute-tools I've described above would be very useful.

So, to make this easier to understand and implement could be to make these tools only available when selecting UI elements that use anchors and offsets for positioning rather than layout containers.

**Comment by golddotasksquestions** [comment]
@lostminds 

Maybe also slightly related (see the "center piece of expanding text" part): 
https://github.com/godotengine/godot/pull/55157#issuecomment-1152968205
https://github.com/godotengine/godot-proposals/issues/129

**Comment by oeleo1** [comment]
Yes, this is a verbatim copy of the original Microsoft Office toolset object ordering facilities which is quite convenient and enjoys a solid user base. However, this does require a container with sub components to be deemed useful wrt the sub component layout, which is an added *magic* to manually craft your margins, expanding or not.

Call me old school, but I do prefer explicit over implicit. Especially when I am programming or designing an interactive interface which will tween the rect_position of your perfectly ordered and aligned items à la Office. So, while I agree this would be a cool goodie, I am not convinced it is a must or even nice to have.

**Reply by lostminds** [reply]
I think this comes down a lot to the workflow you're used to, and I think it's a good thing that Godot offers these two ways of working with layouts as they're both good for different situations. The layout containers feel a lot like a HTML/CSS style system to me, where you build a hierarchy of containers that fully constrain the layout of their sub-containers. In a sense a code/CAD-based approach to layout. And as stated above in that case there's no need for additional aligning tools in the editor, as that's defined by the containers. 

The anchor system however to me feels more suited to a 2D graphic design workflow where you'd move and resize things using the mouse and by setting anchors define what points elements should be positioned relative to if the container is resized. It's when using this type of workflow I think these aligning and distributing tools in the editor would be useful and expected by users. 

I'm not sure I understand what you mean with implicit and explicit definitions or magic in your comment. Perhaps there's some misunderstanding? In my mind these tools would just be another way of changing the position values for controls in the editor. Technically it'd be no different than typing in position values in the inspector or dragging them around in the 2d editor.

**Reply by oeleo1** [reply]
Like I said, agreed it would be another facility in the toolbox to adjust the current anchors and positions, but I (personally) don’t perceive the added value, for two reasons:

One - I don’t really see how this is going to work without some automatically added margin containers to achieve the alignment/distribution requirements (hence I prefer to add/configure them manually Re: explicit vs. implicit magic)

Two - one almost always ends up configuring the container layout in code with the built in presets, especially if you support landscape/portrait switching or other screen adjustments logic, simply for ergonomic reasons. Not to mention the case where the UI panels are dynamic and need to move on/off screen where you are invariably required to code the process, and again based on some built in layout preset.

**Reply by lostminds** [reply]
Yes, regarding your point two as I said I think this depends on how you're used to working with 2d layouts. But it's your point one I don't understand: Why would you need a margin container to be automatically added to move around objects in the 2d editor? Controls that are not placed in layout rule-containers (like centering, margin etc) are just placed based on their transform position, relative to their anchor points in the parent. So moving them around is just a matter of changing their position properties as I wrote above?

**Reply by oeleo1** [reply]
Moving them around - yes. Aligning or distributing them properly - no. You have the position, the size and the pivot offset. So the alignment or equal distribution math isn't that smple in itself for a result you will question most of the time :-)

**Reply by golddotasksquestions** [reply]
@oeleo1 

If you exclude any direct children of Container nodes, what @lostminds wants is perfectly achievable without any added nodes or change of node structure. For both Alignment as well as Distribution.

Godot actually has most of the methods you need to do this built in within the Rect2 class. So with a bit of work you could built this with GDScript (for a plugin for example).

**Reply by oeleo1** [reply]
Okay. I am not saying something smart can’t be done. I am saying that I perceive the added value as low and questionable. That’s what the discussion is about. Throw in a few objects with random size, position, rotation and pivota offsets. Then align and distribute them with the Rect2 tools. I am curious to witness the disappointment. :-). I’d be happy to be proven wrong though!

**Reply by golddotasksquestions** [reply]
The added value is a massive workflow speed up. If you ever created visual layouts and worked with such a tool (they are incredibly common in visual layout and design software) you will know how much of a quality of life improvement they are.

**Reply by oeleo1** [reply]
I do indeed, for static scenarios. Less than 10% of our UI use cases since most of them are dynamic and responsive. But if you perceive the added value as “massive”, go for it. Аlthough this is a shallow point unless 1/ there is a PR to discuss the merits of and 2/ both the added value and quality of life are better quantified. 😊

**Reply by golddotasksquestions** [reply]
Yes, for the most part this is for either static scenarios, or for the design phase during development. Not when you are trying to implement dynamic content on a plethora of different devices.

That being said, even my work on highly dynamic mobile content always had many static sub-arrangements which would greatly benefit from the tools OP proposed here.

**Reply by Hydra9268** [reply]
The Label node's "Anchor preset" drop-down is super handy with its alignment presets. It'd be great if something like that were also available for general objects in the main editor. A quick way to visually align elements would make the 2D layout work much more efficiently. The ideal setup? The core options—Left, Right, and Center for horizontal alignment, and Top, Bottom, and Center for vertical—plus distribute horizontally and vertically. Simple, intuitive, and a huge time-saver.
